### ISSUE TITLE
Getting a connection failure when killing one member of a replica set

### ISSUE THREAD

**Issue by clupprich** [reidmorrison/mongo_ha#1]
Hej Reid!

We've just recently started using mongo_ha in one of our projects and stumbled over an odd case: When we remove one member of a replica set (or kill it), the mongo driver would quit with a Connection Failure message, something that mongo_ha should actually handle, right?

The stack trace is about this:

```
D, [2015-02-12T19:46:31.026136 #14521] DEBUG -- : MONGODB [DEBUG] Logging level is currently :debug which could negatively impact client-side performance. You should set your logging level no lower than :info in production.
I, [2015-02-12T19:46:31.026305 #14521]  INFO -- : MONGODB [INFO] Connecting...
D, [2015-02-12T19:46:31.029509 #14521] DEBUG -- : MONGODB (0.8ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-12T19:46:31.031393 #14521] DEBUG -- : MONGODB (0.6ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-12T19:46:31.032855 #14521] DEBUG -- : MONGODB (0.4ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-12T19:46:31.035125 #14521] DEBUG -- : MONGODB (1.5ms) mongo-replica-test['$cmd'].find({:insert=>"tests", :documents=>[{:timestamp=>2015-02-12 19:46:31 +0100, :_id=>BSON::ObjectId('54dcf507785b1a38b9000001')}], :writeConcern=>{:w=>1}, :ordered=>true}).limit(-1)
D, [2015-02-12T19:46:31.035255 #14521] DEBUG -- : MONGODB (1.8ms) mongo-replica-test['tests'].insert({:timestamp=>2015-02-12 19:46:31 +0100, :_id=>BSON::ObjectId('54dcf507785b1a38b9000001')})
D, [2015-02-12T19:46:32.037136 #14521] DEBUG -- : MONGODB (1.0ms) mongo-replica-test['$cmd'].find({:insert=>"tests", :documents=>[{:timestamp=>2015-02-12 19:46:32 +0100, :_id=>BSON::ObjectId('54dcf508785b1a38b9000002')}], :writeConcern=>{:w=>1}, :ordered=>true}).limit(-1)
D, [2015-02-12T19:46:32.037299 #14521] DEBUG -- : MONGODB (1.4ms) mongo-replica-test['tests'].insert({:timestamp=>2015-02-12 19:46:32 +0100, :_id=>BSON::ObjectId('54dcf508785b1a38b9000002')})
D, [2015-02-12T19:46:33.044079 #14521] DEBUG -- : MONGODB (1.2ms) mongo-replica-test['$cmd'].find({:insert=>"tests", :documents=>[{:timestamp=>2015-02-12 19:46:33 +0100, :_id=>BSON::ObjectId('54dcf509785b1a38b9000003')}], :writeConcern=>{:w=>1}, :ordered=>true}).limit(-1)
D, [2015-02-12T19:46:33.044243 #14521] DEBUG -- : MONGODB (1.6ms) mongo-replica-test['tests'].insert({:timestamp=>2015-02-12 19:46:33 +0100, :_id=>BSON::ObjectId('54dcf509785b1a38b9000003')})
D, [2015-02-12T19:46:34.048153 #14521] DEBUG -- : MONGODB (1.4ms) mongo-replica-test['$cmd'].find({:insert=>"tests", :documents=>[{:timestamp=>2015-02-12 19:46:34 +0100, :_id=>BSON::ObjectId('54dcf50a785b1a38b9000004')}], :writeConcern=>{:w=>1}, :ordered=>true}).limit(-1)
D, [2015-02-12T19:46:34.048317 #14521] DEBUG -- : MONGODB (1.8ms) mongo-replica-test['tests'].insert({:timestamp=>2015-02-12 19:46:34 +0100, :_id=>BSON::ObjectId('54dcf50a785b1a38b9000004')})
D, [2015-02-12T19:46:35.052266 #14521] DEBUG -- : MONGODB (1.1ms) mongo-replica-test['$cmd'].find({:insert=>"tests", :documents=>[{:timestamp=>2015-02-12 19:46:35 +0100, :_id=>BSON::ObjectId('54dcf50b785b1a38b9000005')}], :writeConcern=>{:w=>1}, :ordered=>true}).limit(-1)
D, [2015-02-12T19:46:35.052441 #14521] DEBUG -- : MONGODB (1.6ms) mongo-replica-test['tests'].insert({:timestamp=>2015-02-12 19:46:35 +0100, :_id=>BSON::ObjectId('54dcf50b785b1a38b9000005')})
W, [2015-02-12T19:46:36.053870 #14521]  WARN -- : Connection Failure: 'Operation failed with the following exception: end of file reached' []
I, [2015-02-12T19:46:36.054395 #14521]  INFO -- : MONGODB [INFO] Trying to check replica set health but not connected...
I, [2015-02-12T19:46:36.054503 #14521]  INFO -- : MONGODB [INFO] Initiating hard refresh...
I, [2015-02-12T19:46:36.054607 #14521]  INFO -- : MONGODB [INFO] Connecting...
D, [2015-02-12T19:46:36.057018 #14521] DEBUG -- : MONGODB (0.6ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-12T19:46:36.058924 #14521] DEBUG -- : MONGODB (0.5ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-12T19:46:36.060081 #14521] DEBUG -- : MONGODB [DEBUG] Failed connection to localhost:27002 with Errno::ECONNREFUSED, Connection refused - connect(2) for 127.0.0.1:27002.
W, [2015-02-12T19:46:36.060785 #14521]  WARN -- : Connection Failure: 'Failed to connect to a master node at :' []
/Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo-1.11.1/lib/mongo/mongo_client.rb:422:in `connect': Failed to connect to a master node at : (Mongo::ConnectionFailure)
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo_ha-1.11.1/lib/mongo_ha/mongo_client.rb:101:in `rescue in retry_on_connection_failure'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo_ha-1.11.1/lib/mongo_ha/mongo_client.rb:94:in `retry_on_connection_failure'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo_ha-1.11.1/lib/mongo_ha.rb:35:in `block (3 levels) in <top (required)>'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo-1.11.1/lib/mongo/collection.rb:1091:in `send_write'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo-1.11.1/lib/mongo/collection.rb:419:in `insert'
    from test.rb:25:in `<main>'
```

What I don't understand: why is there a connection failure thrown by MongoClient within the original Mongo driver? Shouldn't this be handled by the monkey-patched versions within mongo_ha? I'm also not sure if I've configured the connection right, this is how it's currently done:

``` ruby
nodes = ['localhost:27001', 'localhost:27002']

options = {
  read:                        :secondary_preferred,
  pool_size:                   100,
  pool_timeout:                10,
  connect_timeout:             5,
  reconnect_attempts:          53,
  reconnect_retry_seconds:     0.1,
  reconnect_retry_multiplier:  2,
  reconnect_max_retry_seconds: 5,
  logger:                      Logger.new(STDOUT)
}

client = Mongo::MongoReplicaSetClient.new(nodes, options)
```

Any hint on what I'm doing wrong?

If you need some more reference - I've extracted a mini project, which you can find in the following gist: https://gist.github.com/clupprich/887d9836fca3a8f19c88

**Comment by reidmorrison** [comment]
Hi Christoph,

Excellent issue submission, you included everything needed to work the issue.

It appears there is a conflict with the latest mongo driver and one of the methods in the patch. I have renamed the retry method and will release an update shortly after testing locally.

Thank you for reporting the issue. We quite likely would have experienced the same issue in our production environment.

Let me know if the updated gem causes any further issues.

Thank you,
Reid

**Comment by clupprich** [comment]
Hi Reid,

Aweoms, thanks for the ultra-fast reply and happy to give valuable and actionable feedback! I've now modified my Gemfile to use the new tag you've created.

I've killed the primary Mongo instance after some inserts and would expect my Mongo connection to failover to the other Mongo instance after the timeout (I've reduced the default values to just 10 reconnect attempts). Now the output looks like that:

```
dione:mongo-replica-test kitto (master)$ ruby test.rb
D, [2015-02-13T08:31:06.834872 #17333] DEBUG -- : MONGODB [DEBUG] Logging level is currently :debug which could negatively impact client-side performance. You should set your logging level no lower than :info in production.
I, [2015-02-13T08:31:06.834991 #17333]  INFO -- : MONGODB [INFO] Connecting...
D, [2015-02-13T08:31:06.838659 #17333] DEBUG -- : MONGODB (0.7ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:06.840389 #17333] DEBUG -- : MONGODB (0.5ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:06.842267 #17333] DEBUG -- : MONGODB (0.6ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:06.845514 #17333] DEBUG -- : MONGODB (2.1ms) mongo-replica-test['$cmd'].find({:insert=>"tests", :documents=>[{:timestamp=>2015-02-13 08:31:06 +0100, :_id=>BSON::ObjectId('54dda83a785b1a43b5000001')}], :writeConcern=>{:w=>1}, :ordered=>true}).limit(-1)
D, [2015-02-13T08:31:06.845638 #17333] DEBUG -- : MONGODB (2.4ms) mongo-replica-test['tests'].insert({:timestamp=>2015-02-13 08:31:06 +0100, :_id=>BSON::ObjectId('54dda83a785b1a43b5000001')})
D, [2015-02-13T08:31:07.847893 #17333] DEBUG -- : MONGODB (1.2ms) mongo-replica-test['$cmd'].find({:insert=>"tests", :documents=>[{:timestamp=>2015-02-13 08:31:07 +0100, :_id=>BSON::ObjectId('54dda83b785b1a43b5000002')}], :writeConcern=>{:w=>1}, :ordered=>true}).limit(-1)
D, [2015-02-13T08:31:07.848075 #17333] DEBUG -- : MONGODB (1.6ms) mongo-replica-test['tests'].insert({:timestamp=>2015-02-13 08:31:07 +0100, :_id=>BSON::ObjectId('54dda83b785b1a43b5000002')})
W, [2015-02-13T08:31:08.851830 #17333]  WARN -- : Connection Failure: 'Operation failed with the following exception: Connection reset by peer' []
D, [2015-02-13T08:31:08.851919 #17333] DEBUG -- : Going to reconnect
I, [2015-02-13T08:31:08.854086 #17333]  INFO -- : MONGODB [INFO] Trying to check replica set health but not connected...
I, [2015-02-13T08:31:08.854195 #17333]  INFO -- : MONGODB [INFO] Initiating hard refresh...
I, [2015-02-13T08:31:08.854261 #17333]  INFO -- : MONGODB [INFO] Connecting...
D, [2015-02-13T08:31:08.857356 #17333] DEBUG -- : MONGODB (0.6ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:08.859500 #17333] DEBUG -- : MONGODB (0.4ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:08.860489 #17333] DEBUG -- : MONGODB [DEBUG] Failed connection to localhost:27002 with Errno::ECONNREFUSED, Connection refused - connect(2) for 127.0.0.1:27002.
W, [2015-02-13T08:31:08.860842 #17333]  WARN -- : Connection unavailable. Waiting: 0.1 seconds before retrying
W, [2015-02-13T08:31:08.964946 #17333]  WARN -- : Connection unavailable. Waiting: 0.2 seconds before retrying
W, [2015-02-13T08:31:09.166939 #17333]  WARN -- : Connection unavailable. Waiting: 0.4 seconds before retrying
W, [2015-02-13T08:31:09.568347 #17333]  WARN -- : Connection unavailable. Waiting: 0.8 seconds before retrying
W, [2015-02-13T08:31:10.371118 #17333]  WARN -- : Connection unavailable. Waiting: 1.6 seconds before retrying
W, [2015-02-13T08:31:11.971732 #17333]  WARN -- : Connection unavailable. Waiting: 3.2 seconds before retrying
W, [2015-02-13T08:31:15.177201 #17333]  WARN -- : Connection unavailable. Waiting: 4.0 seconds before retrying
W, [2015-02-13T08:31:19.180747 #17333]  WARN -- : Connection unavailable. Waiting: 4.0 seconds before retrying
W, [2015-02-13T08:31:23.184973 #17333]  WARN -- : Connection unavailable. Waiting: 4.0 seconds before retrying
W, [2015-02-13T08:31:27.186641 #17333]  WARN -- : Connection unavailable. Waiting: 4.0 seconds before retrying
E, [2015-02-13T08:31:31.191654 #17333] ERROR -- : Auto-reconnect giving up after 10 reconnect attempts
I, [2015-02-13T08:31:31.192009 #17333]  INFO -- : MONGODB [INFO] Trying to check replica set health but not connected...
I, [2015-02-13T08:31:31.192113 #17333]  INFO -- : MONGODB [INFO] Initiating hard refresh...
I, [2015-02-13T08:31:31.192191 #17333]  INFO -- : MONGODB [INFO] Connecting...
D, [2015-02-13T08:31:31.195385 #17333] DEBUG -- : MONGODB (0.8ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:31.197786 #17333] DEBUG -- : MONGODB (0.7ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:31.199006 #17333] DEBUG -- : MONGODB [DEBUG] Failed connection to localhost:27002 with Errno::ECONNREFUSED, Connection refused - connect(2) for 127.0.0.1:27002.
W, [2015-02-13T08:31:31.199408 #17333]  WARN -- : Connection Failure: 'Failed to connect to a master node at :' []
D, [2015-02-13T08:31:31.199498 #17333] DEBUG -- : Going to reconnect
D, [2015-02-13T08:31:31.200055 #17333] DEBUG -- : MONGODB [DEBUG] Checking replica set connection health...
D, [2015-02-13T08:31:31.201776 #17333] DEBUG -- : MONGODB (0.5ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
I, [2015-02-13T08:31:31.202000 #17333]  INFO -- : MONGODB [INFO] Initiating hard refresh...
I, [2015-02-13T08:31:31.202075 #17333]  INFO -- : MONGODB [INFO] Connecting...
D, [2015-02-13T08:31:31.203918 #17333] DEBUG -- : MONGODB (0.4ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:31.205369 #17333] DEBUG -- : MONGODB (0.4ms) admin['$cmd'].find({:ismaster=>1}).limit(-1)
D, [2015-02-13T08:31:31.206247 #17333] DEBUG -- : MONGODB [DEBUG] Failed connection to localhost:27002 with Errno::ECONNREFUSED, Connection refused - connect(2) for 127.0.0.1:27002.
W, [2015-02-13T08:31:31.206536 #17333]  WARN -- : Connection unavailable. Waiting: 0.1 seconds before retrying
W, [2015-02-13T08:31:31.309458 #17333]  WARN -- : Connection unavailable. Waiting: 0.2 seconds before retrying
W, [2015-02-13T08:31:31.513274 #17333]  WARN -- : Connection unavailable. Waiting: 0.4 seconds before retrying
W, [2015-02-13T08:31:31.914583 #17333]  WARN -- : Connection unavailable. Waiting: 0.8 seconds before retrying
W, [2015-02-13T08:31:32.717954 #17333]  WARN -- : Connection unavailable. Waiting: 1.6 seconds before retrying
W, [2015-02-13T08:31:34.323457 #17333]  WARN -- : Connection unavailable. Waiting: 3.2 seconds before retrying
W, [2015-02-13T08:31:37.525941 #17333]  WARN -- : Connection unavailable. Waiting: 4.0 seconds before retrying
W, [2015-02-13T08:31:41.528760 #17333]  WARN -- : Connection unavailable. Waiting: 4.0 seconds before retrying
W, [2015-02-13T08:31:45.529724 #17333]  WARN -- : Connection unavailable. Waiting: 4.0 seconds before retrying
W, [2015-02-13T08:31:49.530024 #17333]  WARN -- : Connection unavailable. Waiting: 4.0 seconds before retrying
E, [2015-02-13T08:31:53.534578 #17333] ERROR -- : Auto-reconnect giving up after 10 reconnect attempts
/Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo-1.11.1/lib/mongo/mongo_client.rb:422:in `connect': Failed to connect to a master node at : (Mongo::ConnectionFailure)
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/bundler/gems/mongo_ha-6cf1f757bbc6/lib/mongo_ha/mongo_client.rb:182:in `block in _reconnect'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/bundler/gems/mongo_ha-6cf1f757bbc6/lib/mongo_ha/mongo_client.rb:161:in `synchronize'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/bundler/gems/mongo_ha-6cf1f757bbc6/lib/mongo_ha/mongo_client.rb:161:in `_reconnect'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/bundler/gems/mongo_ha-6cf1f757bbc6/lib/mongo_ha/mongo_client.rb:116:in `rescue in retry_on_connection_failure'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/bundler/gems/mongo_ha-6cf1f757bbc6/lib/mongo_ha/mongo_client.rb:109:in `retry_on_connection_failure'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/bundler/gems/mongo_ha-6cf1f757bbc6/lib/mongo_ha.rb:28:in `block (3 levels) in <top (required)>'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo-1.11.1/lib/mongo/collection.rb:1091:in `send_write'
    from /Users/kitto/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/mongo-1.11.1/lib/mongo/collection.rb:419:in `insert'
    from test.rb:25:in `<main>'
```

So we get much further than with the last version, but the Mongo driver still says that it cannot find a Mongo host at ":", which kind of looks to me that in the `MongoHA::MongoClient#_reconnect` method, where we call `connect_original`, we expect it to be called on the `MongoReplicaSetClient` but it's actually called on its superclass `MongoClient`.

Thanks!

**Comment by clupprich** [comment]
Hm, I just added a third Mongo instance to the replica set and then the re-connected worked. I haven't looked yet at the internals, but could it be that the Mongo driver automatically falls back to a `MongoClient` instead of a `MongoReplicaSetClient` when only a single Mongo server is left in the replica set? That one is not configured and that why it's failing to access it?

**Comment by reidmorrison** [comment]
With a Mongo Replicaset, if there are only 2 nodes/servers and one goes down the second will switch/stay as secondary. I.e. In order for either one to be master the other has to be running.

In our replicasets we run 2 server nodes and 1 arbiter node to ensure it meets the minimum of 3 nodes during voting.

To see Mongo's behavior tail the Mongo log files while starting and stopping the various nodes. If 2 nodes go down the third will automatically drop down to a secondary, and therefore no master will be available.

I will re-open this ticket in case their is a broader issue.

**Comment by clupprich** [comment]
Thanks Reid! I just had an extensive Friday with all the mongo documentation and a playground open and verified the current behaviour. What's still a bit strange, but that's really, really and edge-case, and actually happening in the original Ruby Mongo driver, is the following case:
- There are only two nodes left in the replicaset, one in the role of secondary and one in primary (I know, three nodes at last, but you know, sometimes,...)
- You're about to have a write operation
- One of the node goes away, the remaining one will become secondary
- Your Mongo stack connects successfully to the last remaining node (secondary)
- You suddenly receive a Socket Error because there's actually no primary node left in the primary pool

I would rather receive something like a NoNodeFoundForOperation error, or something similar, instead of a SocketError, but that's a) not the business of mongo_ha and b) a really edgy case. Also, the ruby mongodriver is going through an extended re-write for the 2.0 release from what I've seen, so things will anyway be quite different.

Thanks again Reid for the quick fix and all the support! Do you know when you'll release 1.11.3 to rubygems?

**Comment by clupprich** [comment]
Just saw that you've already release - thanks!

**Comment by clupprich** [comment]
Hej @reidmorrison! Just wanted to get back to you regarding this issue. Turns out it wasn't solved until we upgraded to the latest version of `mongo_ha`. I suspect that it's this change here that fixes things for the scenario above: https://github.com/reidmorrison/mongo_ha/commit/c22f5295170a4a27bb8d3839249f5586ab678d04#diff-4ac97ebd7eb2ae6d73d964e149049980L54

Up until that point, we always ended up in `Mongo::MongoClient#connect` (instead of `Mongo::MongoReplicaSetClient#connect`) which failed because the Mongo driver wasn't configured for that.
